alicloud/data_source_alicloud_zones.go

```diff
@@ -78,14 +78,26 @@ func dataSourceAlicloudZonesRead(d *schema.ResourceData, meta interface{}) error
 
 	var zoneTypes []ecs.ZoneType
 	for _, types := range resp {
+		if insType == "" && len(types.AvailableInstanceTypes.InstanceTypes) == 0 {
+			continue
+		}
+
 		if insType != "" && !constraints(types.AvailableInstanceTypes.InstanceTypes, insType) {
 			continue
 		}
 
```

Comment:
Line 81 to 87 can be improved by the following expression:

if len(types.AvailableInstanceTypes.InstanceTypes) == 0 ||
          (insType != "" && !constraints(types.AvailableInstanceTypes.InstanceTypes, insType)) {
         continue
}

Same as the next lines.